programs/obfuscator/Obfuscator.cpp

```diff
@@ -1287,32 +1291,57 @@ try
     markov_model_params.determinator_sliding_window_size = options["determinator-sliding-window-size"].as<UInt64>();
 
     /// Create the header block
-    std::vector<std::string> structure_vals;
-    boost::split(structure_vals, structure, boost::algorithm::is_any_of(" ,"), boost::algorithm::token_compress_on);
-
-    if (structure_vals.size() % 2 != 0)
-        throw Exception("Odd number of elements in section structure: must be a list of name type pairs", ErrorCodes::LOGICAL_ERROR);
+    SharedContextHolder shared_context = Context::createShared();
+    auto context = Context::createGlobal(shared_context.get());
+    context->makeGlobalContext();
 
     Block header;
-    const DataTypeFactory & data_type_factory = DataTypeFactory::instance();
 
-    for (size_t i = 0, size = structure_vals.size(); i < size; i += 2)
+    if (structure.empty())
     {
-        ColumnWithTypeAndName column;
-        column.name = structure_vals[i];
-        column.type = data_type_factory.get(structure_vals[i + 1]);
-        column.column = column.type->createColumn();
-        header.insert(std::move(column));
+        ReadBufferIterator read_buffer_iterator = [&]()
+        {
+            return std::make_unique<ReadBufferFromFileDescriptor>(STDIN_FILENO);
+        };
+
+        auto context_const = WithContext(context).getContext();
+
+        auto schema_columns = readSchemaFromFormat(input_format, {}, read_buffer_iterator, false, context_const);
+        auto schema_columns_info = schema_columns.getOrdinary();
+
+        for (auto & info : schema_columns_info)
+        {
+            ColumnWithTypeAndName column;
+            column.name = info.name;
+            column.type = info.type;
+            column.column = column.type->createColumn();
+            header.insert(std::move(column));
+        }
     }
+    else
+    {
+        std::vector<std::string> structure_vals;
+        boost::split(structure_vals, structure, boost::algorithm::is_any_of(" ,"), boost::algorithm::token_compress_on);
 
-    SharedContextHolder shared_context = Context::createShared();
-    auto context = Context::createGlobal(shared_context.get());
-    context->makeGlobalContext();
+        if (structure_vals.size() % 2 != 0)
+            throw Exception("Odd number of elements in section structure: must be a list of name type pairs", ErrorCodes::LOGICAL_ERROR);
+
+        const DataTypeFactory & data_type_factory = DataTypeFactory::instance();
+
+        for (size_t i = 0, size = structure_vals.size(); i < size; i += 2)
+        {
+            ColumnWithTypeAndName column;
+            column.name = structure_vals[i];
+            column.type = data_type_factory.get(structure_vals[i + 1]);
+            column.column = column.type->createColumn();
+            header.insert(std::move(column));
+        }
+    }
```

Comment:
I see that we always had such way to parse table structure, but it's incorrect. Type definition can contain commas as well. For example `t Tuple(x UInt32, y String, ...)`.
We have special function to parse structure from string: `parseColumnsListFromString` in `src/TableFunctions/parseColumnsListFromString.h`